### PR TITLE
Fix for host connection bugs triggered by partitioning test

### PR DESCRIPTION
**Pull request by jumaffre**
Fixes 4 issues triggered by the partition test:
- #2571. More details and related libuv docs (http://docs.libuv.org/en/v1.x/handle.html#c.uv_close): 
> In-progress requests, like uv_connect_t or uv_write_t, are cancelled and have their callbacks called asynchronously with status=UV_ECANCELED.
- Snapshots were enabled, which meant the assumption that a partitioned node would become primary when the partition is lifted wasn't true if a snapshot was generated while the partition was up. 
- Double node connection issue on the host when a follower node would be unable to send responses back to the primary because no node connection was still associated.
- Added a new `--client-connection-timeout-ms` CLI option to `cchost` to specify the TCP timeout for all outgoing TCP connections. This fixes the issue described https://github.com/microsoft/CCF/issues/2616#issuecomment-852253179.

### CODE REVIEW

File: src/host/tcp.h [hunk 1]
```diff
@@ -161,6 +175,29 @@ namespace asynchost
       return service;
     }
 
+    void client_bind()
```
Comment (jumaffre) [inline src/host/tcp.h:178]: Some minor refactor of the client bind as we didn't re-bind to the client interface in the disconnection case.

File: .azure-pipelines-templates/test.yml [hunk 2]
```diff
@@ -5,37 +5,36 @@ parameters:
   suffix: ""
 
 steps:
+  # TODO: Re-enable before merging. Only commented out to get more timeout on partition tests
```
Comment (jumaffre) [inline .azure-pipelines-templates/test.yml:8]: Revert this before merging this PR.

File: src/host/main.cpp [hunk 3]
```diff
@@ -298,6 +295,17 @@ int main(int argc, char** argv)
       "defined by this timer interval.")
     ->capture_default_str();
 
+  size_t client_connection_timeout = 1000;
```
Comment (achamayou) [inline src/host/main.cpp:298]: 1 second seems very aggressive compared to the default (exponential up to 64s)?

File: src/host/node_connections.h [hunk 4]
```diff
@@ -129,9 +126,21 @@ namespace asynchost
...
       {
+        // It is possible that a peer terminates a connection and opens a new
+        // one but the termination is seen by us _after_ messages on the new
+        // connection are received. We re-associate the connection on the latest
+        // received message, so that oubound messages are routed to the correct,
+        // most up-to-date, incoming connection.
+        auto search = parent.associated.find(n);
+        if (search != parent.associated.end() && search->second.first == id)
+        {
+          // Incoming connection is already associated n
```
Comment (achamayou) [inline src/host/node_connections.h:137]:
```suggestion
          // Incoming connection is already associated with n
```

File: src/host/tcp.h [hunk 5]
```diff
@@ -327,14 +360,37 @@ namespace asynchost
...
         LOG_FAIL_FMT("uv_tcp_init failed: {}", uv_strerror(rc));
         return false;
       }
 
+      if (is_client)
+      {
+        uv_os_sock_t sock;
+        if ((sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) == -1)
+        {
+          LOG_FAIL_FMT("socket creation failed: {}", strerror(errno));
```
Comment (achamayou) [inline src/host/tcp.h:376]: If we have the address easily available, it'd be really good to add it to this log line.

**Review by achamayou — APPROVED** [review]

File: src/host/main.cpp (hunk 3 above)
Comment (jumaffre) [reply src/host/main.cpp]: As discussed, this is now 2s so we get at least one retry from the OS, while keeping this value below the default raft election timeout.

File: src/host/tcp.h (hunk 5 above)
Comment (jumaffre) [reply src/host/tcp.h:376]: It's too early at this point as (the address is specified later, on connect.

File: .azure-pipelines-templates/common.yml [hunk 6]
```diff
@@ -6,6 +6,10 @@ jobs:
     ${{ insert }}: ${{ parameters.env }}
 
     steps:
+      - script: |
+          sudo rm -rf build
```
Comment (ashamis) [inline .azure-pipelines-templates/common.yml:10]: This comment could be wrong but should the delete be for . ?

**Review by ashamis — APPROVED** [review]

File: .azure-pipelines-templates/common.yml (hunk 6 above)
Comment (achamayou) [reply .azure-pipelines-templates/common.yml:10]: No, that's the git repo and will be 1. untouched by the tests and 2. reset correctly by DevOps, selectively nuking build/ is the right thing to do.

Comment (jumaffre) [reply .azure-pipelines-templates/common.yml:10]: This cleanup step happens before the checkout so deleting `.` would work too but as @achamayou said, the test doesn't touch anything apart from `build/` (`build/workspace/` in fact) so this should be sufficient.

**Pull request merged**